core/src/contracts/stablex_contract.rs

```diff
@@ -377,6 +367,95 @@ async fn get_block(
         .ok_or_else(|| anyhow!("block {:?} is missing", block_number))
 }
 
+trait BatchIdGetter {
+    fn batch_id_from_block<'a>(&'a self, block_number: BlockNumber) -> BoxFuture<'a, Result<u32>>;
+
+    fn current_batch_id_and_block_number<'a>(&'a self) -> BoxFuture<'a, Result<(u32, u64)>>;
+}
+
+impl BatchIdGetter for Web3<DynTransport> {
+    fn batch_id_from_block<'a>(&'a self, block_number: BlockNumber) -> BoxFuture<'a, Result<u32>> {
+        async move {
+            let current_block = get_block(&self, block_number).await?;
+            Ok(get_block_batch_id(&current_block))
+        }
+        .boxed()
+    }
+
+    fn current_batch_id_and_block_number<'a>(&'a self) -> BoxFuture<'a, Result<(u32, u64)>> {
+        async move {
+            let current_block = get_block(&self, BlockNumber::Latest).await?;
+            let batch_id = get_block_batch_id(&current_block);
+            let block_number = current_block
+                .number
+                .ok_or_else(|| {
+                    anyhow!("latest block {:?} has no block number", current_block.hash)
+                })?
+                .as_u64();
+            Ok((batch_id, block_number))
+        }
+        .boxed()
+    }
+}
+
+trait IncrementalBinarySearch<T: BatchIdGetter = Self> {
+    fn get_last_block_for_batch<'a>(&'a self, batch_id: u32) -> BoxFuture<'a, Result<u64>>;
+}
+
+impl<T: Sync + BatchIdGetter> IncrementalBinarySearch for T {
+    fn get_last_block_for_batch<'a>(&'a self, batch_id: u32) -> BoxFuture<'a, Result<u64>> {
+        async move {
+            struct Bounds {
+                lower: u64,
+                upper: u64,
+            }
+            impl Bounds {
+                fn diff(&self) -> u64 {
+                    self.upper - self.lower
+                }
+                fn mid(&self) -> u64 {
+                    (self.upper + self.lower) / 2
+                }
+            }
+
+            let (current_batch_id, current_block_number) =
+                self.current_batch_id_and_block_number().await?;
+
+            // find lower bound for binary search
+            let mut step = 1_u64;
+            let mut bounds = Bounds {
+                lower: current_block_number,
+                upper: current_block_number,
+            };
+            let mut lower_batch_id = current_batch_id;
+            while batch_id < lower_batch_id {
```

Comment:
Not entirely sure, but I wonder if this might be a prime example of a use case for a `while ... let` statement. I say this because the lower batch id isn't used outside the scope of this loop.

Reply:
I don't think this works. while let is when you have something like an Option that you can match, but in this case the variable and the while condition are separate things.